src/java.base/share/classes/sun/security/ssl/X509Authentication.java

```diff
@@ -204,106 +207,103 @@ private X509PossessionGenerator(String[] keyTypes) {
-                if (SSLLogger.isOn && SSLLogger.isOn("ssl")) {
-                    SSLLogger.finest(
-                            clientAlias + " is not a private key entry");
-                }
-                return null;
+        PrivateKey clientPrivateKey = km.getPrivateKey(clientAlias);
+        if (clientPrivateKey == null) {
+            if (SSLLogger.isOn && SSLLogger.isOn("ssl")) {
+                SSLLogger.finest(
+                        clientAlias + " is not a private key entry");
             }
+            return null;
+        }
 
-            X509Certificate[] clientCerts = km.getCertificateChain(clientAlias);
-            if ((clientCerts == null) || (clientCerts.length == 0)) {
-                if (SSLLogger.isOn && SSLLogger.isOn("ssl")) {
-                    SSLLogger.finest(clientAlias +
+        X509Certificate[] clientCerts = km.getCertificateChain(clientAlias);
+        if ((clientCerts == null) || (clientCerts.length == 0)) {
+            if (SSLLogger.isOn && SSLLogger.isOn("ssl")) {
+                SSLLogger.finest(clientAlias +
                         " is a private key entry with no cert chain stored");
-                }
-                return null;
             }
+            return null;
+        }
 
-            PublicKey clientPublicKey = clientCerts[0].getPublicKey();
-            if ((!clientPrivateKey.getAlgorithm().equals(keyType))
-                    || (!clientPublicKey.getAlgorithm().equals(keyType))) {
-                if (SSLLogger.isOn && SSLLogger.isOn("ssl")) {
-                    SSLLogger.fine(
-                            clientAlias + " private or public key is not of " +
-                            keyType + " algorithm");
-                }
-                return null;
+        PublicKey clientPublicKey = clientCerts[0].getPublicKey();
+        if (!clientPrivateKey.getAlgorithm().equals(clientPublicKey.getAlgorithm())) {
+            if (SSLLogger.isOn && SSLLogger.isOn("ssl")) {
+                SSLLogger.fine(
+                        clientAlias + " private or public key is not of " +
+                                "same algorithm: " +
+                                clientPrivateKey.getAlgorithm() + " vs " +
+                                clientPublicKey.getAlgorithm());
             }
-
-            return new X509Possession(clientPrivateKey, clientCerts);
+            return null;
         }
 
-        private SSLPossession createServerPossession(
-                ServerHandshakeContext shc, String keyType) {
-            X509ExtendedKeyManager km = shc.sslContext.getX509KeyManager();
-            String serverAlias = null;
+        return new X509Possession(clientPrivateKey, clientCerts);
+    }
+
+    private static SSLPossession createServerPossession(
+            ServerHandshakeContext shc, String[] keyTypes) {
+        X509ExtendedKeyManager km = shc.sslContext.getX509KeyManager();
+        String serverAlias = null;
+        for (String keyType : keyTypes) {
```

Comment:
What do you think if we update the createServerPossession to call chooseServerAlias only once?  A similar problem could occur in server side, I think.  Keeping the behavior consistent between client and server may easy the key manager development and customization.

Reply:
Do not fully understand. `chooseServerAlias` can only take one key type. How do I can it only once?

Reply:
Never mind, I did not realized only one key type accepted.  Hm, the for-loop is a little bit weird to me now.  Only one keyType should be initialized for server.  Anyway, not a big concern and you can leave is as is.